### ISSUE TITLE
Extend Communication Data Length Limitation

### ISSUE THREAD

**Issue by yim0331** [APPFL/APPFL#112]
Extend MPI communication data length limitation by leveraging BigMPI4py: https://gitlab.com/alexmascension/bigmpi4py

This issue is generated by the problem found in issue #108.

The problem is the MPI collective function does not support parallelization for data that is greater than 2^31 bytes. For the ResNet model with the LEAF dataset, APPFL faces the limitation while gathering the data from all clients.

**Comment by yim0331** [comment]
Update:
Although using bigmpi4py, APPFL encounters the same issue of the data length limitation:
Traceback (most recent call last):
  File "./femnist.py", line 161, in <module>
    main()
  File "./femnist.py", line 154, in main
    rm.run_client(cfg, comm, model, loss_fn, args.num_clients, train_datasets)
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/run_mpi.py", line 287, in run_client
    BM.gather(local_states, comm, root=0)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/bigmpi4py/bigmpi4py.py", line 1481, in gather
    return _general_gather(gather_object, comm,  size_limit, root, optimize=optimize, gather_method='gather')
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/bigmpi4py/bigmpi4py.py", line 1235, in _general_gather
    object_return = _gather_or_allgather(gather_object, comm, root, gather_method, optimize=optimize)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/bigmpi4py/bigmpi4py.py", line 690, in _gather_or_allgather
    return _gatherv(object, comm, root, optimize, k1_val)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/bigmpi4py/bigmpi4py.py", line 685, in _gatherv
    return comm.gather(object, root=root)
  File "mpi4py/MPI/Comm.pyx", line 1578, in mpi4py.MPI.Comm.gather
  File "mpi4py/MPI/msgpickle.pxi", line 772, in mpi4py.MPI.PyMPI_gather
  File "mpi4py/MPI/msgpickle.pxi", line 147, in mpi4py.MPI.pickle_dump
  File "mpi4py/MPI/msgbuffer.pxi", line 50, in mpi4py.MPI.downcast
OverflowError: integer 8842650051 does not fit in 'int'

It seems the reason is the data structure of send buffers. APPFL uses OrderedDict() object to organize send buffers. The dictionary object seems not to be able to be partitioned in BigMPI4Py because the data in a key consider a chunk.  So, I'm thinking to convert the data structure to the list vectors.
However, I am concerned about the side effect of the changes in a data structure. @minseok-ryu what do you think of this idea?

**Comment by kibaekkim** [comment]
Looking at the message, it looks like the code is still using `mpi4py`. Please make sure `bigmpi4py` is used.

**Comment by yim0331** [comment]
Update:
While converting the Gather function to the manual send-recv function, The Extremely Slow Warning and Out of Memory Error are rising on the SWING GPU cluster machine.

```
./femnist.py:75: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  /opt/conda/conda-bld/pytorch_1659484810403/work/torch/csrc/utils/tensor_new.cpp:201.)
  torch.FloatTensor(test_data_input), torch.tensor(test_data_label)
/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.
  warnings.warn(
/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)
Traceback (most recent call last):
  File "./femnist.py", line 161, in <module>
    main()
  File "./femnist.py", line 152, in main
    rm.run_server(cfg, comm, model, loss_fn, args.num_clients, test_dataset, args.dataset)
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/run_mpi.py", line 134, in run_server
    local_states[cid] = comm.recv(source=rank, tag=cid)
  File "mpi4py/MPI/Comm.pyx", line 1438, in mpi4py.MPI.Comm.recv
  File "mpi4py/MPI/msgpickle.pxi", line 341, in mpi4py.MPI.PyMPI_recv
  File "mpi4py/MPI/msgpickle.pxi", line 306, in mpi4py.MPI.PyMPI_recv_match
  File "mpi4py/MPI/msgpickle.pxi", line 152, in mpi4py.MPI.pickle_load
  File "mpi4py/MPI/msgpickle.pxi", line 141, in mpi4py.MPI.cloads
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/storage.py", line 222, in _load_from_bytes
    return torch.load(io.BytesIO(b))
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/serialization.py", line 713, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/serialization.py", line 930, in _legacy_load
    result = unpickler.load()
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/serialization.py", line 876, in persistent_load
    wrap_storage=restore_location(obj, location),
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/serialization.py", line 175, in default_restore_location
    result = fn(storage, location)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/serialization.py", line 155, in _cuda_deserialize
    return torch._UntypedStorage(obj.nbytes(), device=torch.device(location))
RuntimeError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 39.59 GiB total capacity; 7.09 GiB already allocated; 5.19 MiB free; 7.45 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
./femnist.py:75: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  /opt/conda/conda-bld/pytorch_1659484810403/work/torch/csrc/utils/tensor_new.cpp:201.)
  torch.FloatTensor(test_data_input), torch.tensor(test_data_label)
/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.
  warnings.warn(
/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)
Traceback (most recent call last):
  File "./femnist.py", line 161, in <module>
    main()
  File "./femnist.py", line 154, in main
    rm.run_client(cfg, comm, model, loss_fn, args.num_clients, train_datasets)
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/run_mpi.py", line 287, in run_client
    local_states[cid] = client.update()
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/algorithm/client_optimizer.py", line 65, in update
    loss.backward()
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/_tensor.py", line 396, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: Unable to find a valid cuDNN algorithm to run convolution
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[45757,1],0]
  Exit code:    1
--------------------------------------------------------------------------
```

To minimize the communication buffer, the current MPI communication implementation transfers data for each 'cid' with  blocking calls:

Server Implementation:
```
local_states={}
        for rank in range(comm_size):
            if rank == 0:
                continue;
            else:
                for _, cid in enumerate(num_client_groups[rank - 1]):
                    local_states[cid] = comm.recv(source=rank, tag=cid) 
```
Client Implementation:
```
local_states = {}

    while do_continue:
        """Receive "global_state" """
        global_state = comm.bcast(None, root=0)

        """ Update "local_states" based on "global_state" """
        for client in clients:
            cid = client.id
            ## initial point for a client model            
            client.model.load_state_dict(global_state)

            ## client update
            local_states[cid] = client.update()      
            req = comm.send(local_states[cid], 0, tag=cid)  
```

**Comment by yim0331** [comment]
Update:
After copying the local_states from GPU memory to CPU memory in client processors before MPI communication, the error, RuntimeError: CUDA out of memory, was resolved. 

However, another error occurs:
```
./femnist.py:75: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  /opt/conda/conda-bld/pytorch_1659484810403/work/torch/csrc/utils/tensor_new.cpp:201.)
  torch.FloatTensor(test_data_input), torch.tensor(test_data_label)
/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.
  warnings.warn(
/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)
Traceback (most recent call last):
  File "./femnist.py", line 161, in <module>
    main()
  File "./femnist.py", line 154, in main
    rm.run_client(cfg, comm, model, loss_fn, args.num_clients, train_datasets)
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/run_mpi.py", line 291, in run_client
    ls = client.update()
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/algorithm/client_optimizer.py", line 63, in update
    output = self.model(data)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/gpfs/fs1/home/yim/workspace/appfl-benchmark/FEMNIST_Sample/models/resnet.py", line 98, in forward
    x = self.layer4(x)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/container.py", line 139, in forward
    input = module(input)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/gpfs/fs1/home/yim/workspace/appfl-benchmark/FEMNIST_Sample/models/resnet.py", line 31, in forward
    out = self.conv2(out)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 457, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 453, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: Unable to find a valid cuDNN algorithm to run convolution
```

The error occurs while 168's client.update(). In the testing logs, I could check the 167 times client.update() was completed before the error. The total number of clients is 197 in the data.
Currently, I'm suspecting GPU memory. This is because the 168's client update was the point of previously getting the error, "CUDA out of memory" with the message, "7.09 GiB already allocated; 5.19 MiB free".

**Comment by yim0331** [comment]
Update:
After the investigation, I found that the issue, RuntimeError: Unable to find a valid cuDNN algorithm to run convolution occurred due to the out of GPU memory. To handle this issue, the primal tensor variables in the model state in each client moved to CPU memory. Also, some unnecessary deepcopy function calls were removed. With these changes, I could check that all client sends their model status to the server successfully. 

However, the CUDA out-of-memory issue occurs during updating primal variables for server updates.
```
Traceback (most recent call last):
  File "./femnist.py", line 161, in <module>
    main()
  File "./femnist.py", line 152, in main
    rm.run_server(cfg, comm, model, loss_fn, args.num_clients, test_dataset, args.dataset)
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/run_mpi.py", line 139, in run_server
    server.update(local_states)
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/algorithm/server_federated.py", line 64, in update
    super(FedServer, self).primal_recover_from_local_states(local_states)
  File "/gpfs/fs1/home/yim/workspace/APPFL-install/APPFL/src/appfl/algorithm/algorithm.py", line 62, in primal_recover_from_local_states
    self.primal_states[sid] = copy.deepcopy(states["primal"])
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/copy.py", line 296, in _reconstruct
    value = deepcopy(value, memo)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/copy.py", line 153, in deepcopy
    y = copier(memo)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/_tensor.py", line 110, in __deepcopy__
    new_storage = self.storage().__deepcopy__(memo)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/storage.py", line 569, in __deepcopy__
    return self._new_wrapped_storage(copy.deepcopy(self._storage, memo))
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/copy.py", line 153, in deepcopy
    y = copier(memo)
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/storage.py", line 89, in __deepcopy__
    new_storage = self.clone()
  File "/home/yim/.conda/envs/APPFL/lib/python3.8/site-packages/torch/storage.py", line 103, in clone
    return type(self)(self.nbytes(), device=self.device).copy_(self)
RuntimeError: CUDA out of memory. Tried to allocate 2.00 MiB (GPU 0; 39.59 GiB total capacity; 10.11 GiB already allocated; 3.19 MiB free; 10.63 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

I'm investigating the cause of the issue. In my understanding, the server side uses only CPU, so the CUDA out-of-memory error should not happen.  The latest update code is committed to [FEMNIST_example branch](https://github.com/APPFL/APPFL/commit/7e4c5ee1aaa0fbb79a032b3c1dbbc7db6148e71c)

I will keep updating this post, but also @Zilinghan please feel free to contact me with any questions.

**Comment by kibaekkim** [comment]
@yim0331 The server is supposed to run on CPU. Can you check that?

**Comment by yim0331** [comment]
@kibaekkim I'm investigating that part. 
In my test setup, I could check the server instance is created by FedServer class with device = 'cpu'. Also, the model in the server is set to use the 'cpu'. 
I'm currently suspecting the deepcopy function that supposes to assign the copied variables on the cpu memory, but it seems it assigns the copied variables on CUDA memory.

**Comment by kibaekkim** [comment]
@yim0331 Can you check every code line to see where the memory is created?

**Comment by yim0331** [comment]
Fixed by #116
